src/Storage/Notmuch.hs

```diff
@@ -226,30 +229,34 @@ getThreads
   :: (MonadError Error m, MonadIO m)
   => T.Text
   -> NotmuchSettings FilePath
-  -> m (V NotmuchThread)
+  -> m (V (Toggleable NotmuchThread))
 getThreads s settings =
   withDatabaseReadOnly (view nmDatabase settings) $
     flip Notmuch.query (Notmuch.Bare $ T.unpack s)
     >=> Notmuch.threads
-    >=> liftIO . fmap (fromList 128) . lazyTraverse threadToThread
+    >=> liftIO . fmap (fromList 128) . lazyTraverse (threadToThread >=> \t -> pure (False, t))
 
 lazyTraverse :: (a -> IO b) -> [a] -> IO [b]
 lazyTraverse f =
   foldr (\x ys -> (:) <$> f x <*> unsafeInterleaveIO ys) (pure [])
 
--- | returns a vector of *all* messages belonging to the given thread
+-- | Returns a vector of *all* messages belonging to the list of threads
 --
 getThreadMessages
-  :: (MonadError Error m, MonadIO m)
+  :: (MonadError Error m, MonadIO m, Traversable t)
   => FilePath
-  -> NotmuchThread
-  -> m (Vec.Vector NotmuchMail)
-getThreadMessages fp t =
-  withDatabaseReadOnly fp go
-  where go db = do
-          msgs <- getThread db (view thId t) >>= Notmuch.messages
-          mails <- liftIO $ traverse messageToMail msgs
-          pure $ Vec.fromList mails
+  -> t NotmuchThread
+  -> m (Vec.Vector (Toggleable NotmuchMail))
+getThreadMessages fp ts = withDatabaseReadOnly fp go
+  where
+    go db = do
+      msgs <- foldlM (allMessages db) [] ts
```

Comment:
Why `foldlM` specifically?

Would something like
```haskell
msgs <- mconcat <$> traverse (Notmuch.messages <=< getThread db . view thId) ts
```

work?  I just find traversals and monoids much easier to read than folds.

*edit* you could even avoid the `mconcat` and just do another layer of `traverse` in the next line, e.g.:
```haskell
msgs <- traverse (Notmuch.messages <=< getThread db . view thId) ts
mails <- liftIO $ (traverse . traverse) messageToMail msgs
```

Reply:
You know I think this is a remnant of me fiddling around. I'll have a look and change. Cheers!

Reply:
Hmm... so I looked into this and the problem is that `traverse` seems not to be the right abstraction. With `traverse` I get:

```
t [Notmuch.Message 0 mode]
```
as my `msgs`, while with `foldM` I get my list of messages returned. I've checked our notmuch binding which has:

```Haskell
class HasMessages a where
  messages
    :: (AsNotmuchError e, MonadError e m, MonadIO m)
    => a mode -> m [Message 0 mode]
```
In a way it does make sense to me (not sure if I get this right), since `traverse` is defined over `Applicative f`, while our monadic list would be one level deeper since `Monad` is a subclass of `Applicative`.

Yes?

Reply:
So I slept over this a night and I think I got it backwards. It should work since traverse already works for `Applicative` and since `Monad` is a subclass of `Applicative` it should also work for `Monad`... hrm

Reply:
@romanofski see https://github.com/purebred-mua/purebred/commit/9b0e18728b84bd25de7df01fcc21d2e6b0cf77e1.  For your consideration (i.e. it's not a showstopper; if you prefer `foldM` we'll keep it as-is).

Reply:
Hah no.. that looks good. I'll have to keep this in mind!